src/report/ReportPoller.ts

```diff
@@ -0,0 +1,138 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { Mjolnir, REPORT_POLL_EVENT_TYPE } from "../Mjolnir";
+import { ReportManager } from './ReportManager';
+import { LogLevel } from "matrix-bot-sdk";
+
+class InvalidStateError extends Error {}
+
+export class ReportPoller {
+    /*
+     * https://matrix-org.github.io/synapse/latest/admin_api/event_reports.html
+     * "from" is an opaque token that is returned from the API to paginate reports
+     */
+    private from = 0;
+    private timeout: ReturnType<typeof setTimeout> | null = null;
+
+    /**
+     * A class to poll synapse's report endpoint, so we can act on new reports
+     *
+     * @param mjolnir The running Mjolnir instance
+     * @param manager The report manager in to which we feed new reports
+     */
+    constructor(
+        private mjolnir: Mjolnir,
+        private manager: ReportManager,
+    ) { }
+
+    private schedulePoll() {
+        if (this.timeout === null) {
+            /*
+             * Important that we use `setTimeout` here, not `setInterval`,
+             * because if there's networking problems and `getAbuseReports`
+             * hangs for longer thank the interval, it could cause a stampede
+             * of requests when networking problems resolve
+             */
+            this.timeout = setTimeout(
+                this.tryGetAbuseReports.bind(this),
+                60_000 // a minute in milliseconds
+            );
+        } else {
+            throw new InvalidStateError("poll already scheduled");
+        }
+    }
+
+    private async getAbuseReports() {
+        let response_: {
+            event_reports: { room_id: string, event_id: string, sender: string, reason: string }[],
+            next_token: number | undefined
+        } | undefined;
+        try {
+            response_ = await this.mjolnir.client.doRequest(
+                "GET",
+                "/_synapse/admin/v1/event_reports",
+                { from: this.from.toString() }
+            );
+        } catch (ex) {
+            await this.mjolnir.logMessage(LogLevel.ERROR, "getAbuseReports", `failed to poll events: ${ex}`);
+            this.schedulePoll();
+            return;
+        }
+
+        const response = response_!;
+        for (let report of response.event_reports) {
+            let event: any; // `any` because `handleServerAbuseReport` uses `any`
+            try {
+                event = (await this.mjolnir.client.doRequest(
+                    "GET",
+                    `/_synapse/admin/v1/rooms/${report.room_id}/context/${report.event_id}?limit=1`
+                )).event;
+            } catch (ex) {
+                continue;
```

Comment:
I assume we're not logging because there is a chance that we may not be in the rooms, right? But if so, we may end up losing some info.

Also, we'll need to check the performance of this. If we perform requests for rooms we're not part of, this could end up expensive.

How complicated would it be to check the list of rooms we're in and only iterating on the rooms we're a member of?

If it's too complex, we can keep this as a followup.